README.md
# ELS - Enterprise Linux/EuroLinux Sources - system-switch-java
 
## Usage:
  Checkout branch or tag. Use `make sources` to recreate  SOURCE0 archive.


src/ChangeLog
2016-08-07  Jiri Vanek <jvanek@redhat.com>

	improved tui composing code around close and new cancel button

	* switch_java_functions.py: added Cancel message
	* switch_java_tui.py: buttons moved to groups

2016-08-07  Jiri Vanek <jvanek@redhat.com>

	hardcoded usage of python2
	* system-switch-java.in: moved to use /usr/bin/python2

2016-08-07  Jiri Vanek <jvanek@redhat.com>

	Added switch -t to allow walk around hardcoded check on UID 0

	* switch_java_boot.py: added new option '-r', '--non-root' whci set
	switch_java_functions.none_root to  True
	(main) check for UID softened by usage of none_root
	* switch_java_functions.py: added new field none_root, by default False	

2016-08-07  Jiri Vanek <jvanek@redhat.com>

	Added support for zipped docs and java-version-vendor masters.
	Fixed call to crating javadoc master by adding /api suffix
	* switch_java_functions.py: new dictionary JAVADOCZIP.
	(_get_javadocs_alternatives) new method  to unify getting javadoc and zipped docs
	(get_javadocdir_alternatives) and (get_javadoczip_alternatives) wrrapping
	_get_javadocs_alternatives by .zip or /api suffix.
	MEthods used reasonably.
	

2016-03-23  Omair Majid  <omajid@redhat.com>

	* po/system-switch-java.pot: Regenerated.
	* switch_java_functions.py (get_java_identifiers)
	(initialize_alternatives_dictionaries): Accept JDK directories without a
	'jre' sub-directory.
	* tests/test_switch_java_functions.py:
	(test_get_java_identifiers_finds_jre_in_jdk_without_jre_subdir)
	(test_initialize_alternative_dictionaries_picks_up_java_with_jre_subdir)
	(test_initialize_alternative_dictionaries_picks_up_java_without_jre_subdir)
	(test_initialize_alternative_dictionaries_picks_up_jre_without_jre_subdir):
	New tests.

2015-11-23  Jiri Vanek <jvanek@redhat.com>

	Added support for family flag of future alternatives
	* switch_java_functions.py: (_get_alternatives) when parsing priority
	it checks if new format with @ delimiters is presented. If so, parse it, if no
	direct convert to int is done as before. If debug, this method prints out why
	it died on exception before rethrowing

2015-11-20  Jiri Vanek <jvanek@redhat.com>

	Added support for debug subpackages
	* switch_java_functions.py: (get_pretty_name) now strip -debug wen generating
	pretty name and addinng DEBUG if -debug was -debug was presented in NVRA
	(get_java_split) also strips -debug from NVRA
	* switch_java_tui.py: tui dialog made 10 chars more wide.	

2015-11-20  Jiri Vanek <jvanek@redhat.com>

	Added ITW support
	* switch_java_functions.py: (get_java_identifiers) added debug output of read
	alternatives. (get_plugin_alternatives) regex enhanced for IcedTeaPlugin.so
	(initialize_alternatives_dictionaries) made aware of IcedTeaPlugin.so too.

2015-11-20  Jiri Vanek <jvanek@redhat.com>

	Added debug switch
	* switch_java_boot.py: made aware of switch_java_functions.ssj_debug. Added
	-d/--debug option to be known and used.
	* switch_java_functions.py: declared ssj_debug with False. (_exec) id debug
	then it prints out command it is executing
	* switch_java_globals.py.in: removed comment of replacing prefix by bindir
	and friends.Its working and not going to be fixed any time soon

2014-04-25  Omair Majid  <omajid@redhat.com>

	* switch_java_functions.py: Style fixes. Remove trailing semicolons,
	parentheses in if statements and extra or missing spacing.

2014-04-25  Omair Majid  <omajid@redhat.com>

	* tests/test_switch_java_functions.py: Disable some pylint warnings.

2014-04-25  Omair Majid  <omajid@redhat.com>

	* switch_java_functions.py
	(get_alternatives): Add optional arguments for unit testing. Wrap
	file operations in a with block.
	(_get_alternatives): New function.
	* tests/test_switch_java_functions.py
	(file_always_exists, create_stub_open_that_raises)
	(create_stub_open_for): New functions.
	(test_get_alternatives): Remove.
	(test_get_alternatives_raises_JavaOpenError_on_missing_alternative_file)
	(test_get_alternatives_raises_JavaParseError_on_empty_file)
	(test_get_alternatives_empty_alternatives_best_is_negative)
	(test_get_alternatives_parses_alternatives)
	(test_get_alternatives_selects_best_alternatives): New methods.

2014-04-24  Omair Majid  <omajid@redhat.com>
	* tests/test_switch_java_functions.py
	(test_switch_java_changes_java_and_jre_alternatives)
	(test_switch_java_changes_jce_alternatives)
	(test_switch_java_changes_javac_alternatives)
	(test_switch_java_changes_plugin_alternatives)
	(test_switch_java_changes_javadocdir_alternatives): Use self.assertIn for
	lists.

2014-04-24  Omair Majid  <omajid@redhat.com>

	* switch_java_functions.py
	(switch_java): Call _set_alternative instead of building command line
	every time.
	(_set_alternative): New function.

2014-04-24  Omair Majid  <omajid@redhat.com>

	* switch_java_functions.py
	(_exec): New function.
	(switch_java): Call _exec instead of os.system.
	* tests/test_switch_java_functions.py
	(test_switch_java): Remove.
	(__init__, setUp, tearDown, _reset_globals)
	(test_switch_java_changes_java_and_jre_alternatives)
	(test_switch_java_changes_jce_alternatives)
	(test_switch_java_changes_javac_alternatives)
	(test_switch_java_changes_plugin_alternatives)
	(test_switch_java_changes_javadocdir_alternatives)
	(_populate_empty_globals): New functions.

2014-04-24  Omair Majid  <omajid@redhat.com>

	* switch_java_functions.py: Add module-level docstring. Remove uneeded
	imports
	(JavaOpenError): Add docstring.
	(JavaParseError): Add docstring.
	(get_java_identifiers): Fix typo in variable name.
	(get_alternatives): Don't use names that clash with built-in functions.
	* switch_java_tui.py: Add module-level docstring.
	(main): Don't use names that clash with built-in types.
	* tests/test_switch_java_functions.py: Add module-level docstring.
	(test_get_java_split_with_arch, test_get_java_split_without_arch): Don't
	use names that clash with built-in functions.
	* po/system-switch-java.pot: Regenerate.

2014-04-23  Jiri Vanek  <jvanek@redhat.com>


src/switch_java_gui.py
# switch_java_gui - the Java toolset switcher GUI
# Copyright (C) 2007 Red Hat, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA.

import gtk
import gtk.glade
from switch_java_functions import *
from switch_java_globals import DATA_ROOT_DIR, PACKAGE_NAME

class mainDialog:
    def __init__(self):
        pass

    def main(self, java_identifiers, default_java, pretty_names):
        # Initialize UI.
        xml = gtk.glade.XML(DATA_ROOT_DIR + '/' + PACKAGE_NAME + '/' +
                            PACKAGE_NAME + '.glade',
                            None, domain=PACKAGE_NAME)
        self.dialog = xml.get_widget('dialog')
        self.dialog.set_title(TITLE_MESSAGE)
        radio_vbox = xml.get_widget('radio-vbox')
        xml.get_widget('instruction-label').set_text(INSTRUCTION_MESSAGE)
        xml.get_widget('selection-label').set_markup('<b>'
                                                     + SELECTION_MESSAGE
                                                     + '</b>')
        ok_button = xml.get_widget('ok-button')
        # Initialize alternatives list.
        self.radio_buttons = {}
        self.alternatives = {}
        for java in java_identifiers:
            if len(self.radio_buttons) == 0:
                self.radio_buttons[java] = gtk.RadioButton(None,
                                                           pretty_names[java])
                group_button = self.radio_buttons[java]
            else:
                self.radio_buttons[java] = gtk.RadioButton(group_button,
                                                           pretty_names[java])
            radio_vbox.pack_start(self.radio_buttons[java], False, False, 0)
            self.radio_buttons[java].set_use_underline(False);
            self.radio_buttons[java].show()
            self.radio_buttons[java].connect("toggled", lambda *x:
                                             ok_button.set_sensitive(True))
            self.alternatives[self.radio_buttons[java]] = java
        # Show default alternative as currently selected and give it
        # keyboard focus.
        self.radio_buttons[default_java].set_active(True)
        self.radio_buttons[default_java].grab_focus()
        ok_button.connect('clicked', self.ok_button_clicked)
        ok_button.set_sensitive(False)
        xml.get_widget('cancel-button').connect('clicked',
                                                self.cancel_button_clicked)
        self.dialog.connect('delete-event', self.dialog_delete_event)
        self.dialog.connect('hide', gtk.main_quit)
        self.dialog.set_icon_from_file(DATA_ROOT_DIR + '/pixmaps/'
                                       + PACKAGE_NAME + '.png')
        self.dialog.show()
        gtk.main()

    def dialog_delete_event(self, *args):
        gtk.main_quit()

    def ok_button_clicked(self, button):
        for radio_button in self.radio_buttons.values():
            if radio_button.get_active():
                switch_java(self.alternatives[radio_button])
                break
        self.dialog.hide()
        gtk.main_quit()

    def cancel_button_clicked(self, button):
        self.dialog.hide()
        gtk.main_quit()

    def show_dialog(self, message):
        dialog = gtk.MessageDialog(None,
                                   gtk.DIALOG_MODAL
                                   | gtk.DIALOG_DESTROY_WITH_PARENT,
                                   gtk.MESSAGE_INFO,
                                   gtk.BUTTONS_OK,
                                   message)
        dialog.set_title(TITLE_MESSAGE)
        dialog.set_icon_from_file(DATA_ROOT_DIR + '/pixmaps/' + PACKAGE_NAME + '.png')
        dialog.run()
        dialog.destroy()


src/switch_java_tui.py
# switch_java_tui - the Java toolset switcher TUI
# Copyright (C) 2007 Red Hat, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA.

'''The text UI for switching the java toolet'''

from snack import *
from switch_java_functions import *

class mainDialog:
    def __init__(self):
        pass

    def main(self, java_identifiers, default_java, pretty_names):
        # Initialize UI.
        screen = SnackScreen()
        text = TextboxReflowed(60, INSTRUCTION_MESSAGE)
        java_list = Listbox(5, width=60, scroll=1, returnExit=1)
        label = Label(SELECTION_MESSAGE)
        # Initialize alternatives list.
        for java in java_identifiers:
            java_list.append(pretty_names[java], java)
        # Show default alternative as currently selected and give it
        # keyboard focus.
        java_list.setCurrent(default_java)
        buttons = (CLOSE_MESSAGE, CANCEL_MESSAGE)
        buttonbar = ButtonBar(screen,buttons)
        grid = GridForm(screen, TITLE_MESSAGE, 1, 5)
        grid.add(text, 0, 0)
        grid.add(label, 0, 1, (0, 1, 0, 0))
        grid.add(java_list, 0, 2, (0, 1, 0, 0))
        grid.add(buttonbar, 0, 3, (0, 1, 0, 0))
        result = grid.runOnce()
        screen.finish()
        bbp = buttonbar.buttonPressed(result)
        if ssj_debug:
            print("  bbpr    " + str(bbp))
            print("  message " + CANCEL_MESSAGE)
            print("  result  " + str(result))
        if bbp == "cancel":
            print("Not saving!")
        else:
            switch_java(java_list.current())

    def show_dialog(self, message):
        screen = SnackScreen()
        button = Button(OK_MESSAGE)
        text = TextboxReflowed(40, message + '\n\n')
        grid = GridFormHelp(screen, TITLE_MESSAGE, None, 1, 2)
        grid.add(text, 0, 0)
        grid.add(button, 0, 1)
        result = grid.runOnce()
        screen.finish()


src/switch_java_boot.py
# switch_java_boot - the Java toolset switcher frontend
# Copyright (C) 2007 Red Hat, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA.

import os
import os.path
import sys
from optparse import OptionParser

from switch_java_globals import VERSION
import switch_java_functions
from switch_java_functions import *

def main():
    parser = OptionParser(version=VERSION)
    parser.add_option('-t', '--text', action='store_true',
                      dest='force_text', default=False, help=TEXT_MESSAGE)
    parser.add_option('-d', '--debug', action='store_true',
                      dest='force_debug', default=False, help="")
    parser.add_option('-r', '--non-root', action='store_true',
                      dest='none_root', default=False, help="")
    options, args = parser.parse_args()
    show_gui = False
    if options.force_debug:
        switch_java_functions.ssj_debug=True
    if options.none_root:
        switch_java_functions.none_root = True
    if not options.force_text:
        try:
            if os.environ['DISPLAY'] != '':
                show_gui = True
        except KeyError:
            pass
    if os.getuid() != 0 and not switch_java_functions.none_root:
        show_dialog(show_gui, ROOT_MESSAGE)
        sys.exit(1)
    java_identifiers = []
    best_identifier = ''
    # Get list of identifiers for installed Java environments.
    try:
        java_identifiers, best_identifier = get_java_identifiers()
    except JavaParseError:
        show_dialog(show_gui, PARSE_ERROR_MESSAGE)
        sys.exit(1)
    except JavaOpenError:
        pass
    if len(java_identifiers) == 0:
        show_dialog(show_gui, NO_JAVA_MESSAGE)
        sys.exit(1)
    # Get default Java alternative.  Default to best alternative if no
    # recognized default is currently set.  That is, if
    # /etc/alternatives/java does not point to a recognized Java
    # alternative, point it to the best alternative.  This will
    # override a deliberate custom setting of /etc/alternatives/java
    # but that's probably the right thing to do for an easy-to-use GUI
    # tool.  In almost all cases /etc/alternatives/java pointing to
    # something unrecognized will represent an error condition that
    # should be fixed, rather than a custom setting.
    default_java_command = get_default_java_command()
    if default_java_command not in JAVA.values():
        default_java_command = JAVA[best_identifier]
    default_java = ALTERNATIVES[default_java_command]
    pretty_names = get_pretty_names(java_identifiers)
    if show_gui:
        from switch_java_gui import mainDialog
    else:
        from switch_java_tui import mainDialog
    mainDialog().main(java_identifiers, default_java, pretty_names)

def show_dialog(show_gui, message):
    if show_gui:
        from switch_java_gui import mainDialog
    else:
        from switch_java_tui import mainDialog
    mainDialog().show_dialog(message)


src/switch_java_globals.py
# switch_java_globals - global constants
# Copyright (C) 2014 Red Hat, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA.

PACKAGE_NAME='system-switch-java'
PROGNAME='system-switch-java'
VERSION='1.1.7.1'

PREFIX='/usr/local'
BIN_DIR='/usr/local/bin'
DATA_ROOT_DIR='/usr/local/share'
LOCALE_DIR='/usr/local/locale'


src/switch_java_functions.py
# switch_java_functions - functions for switching Java alternatives
# Copyright (C) 2007 Red Hat, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA.

'''Functions for manipulating the java alternatives.'''

import gettext
import os
import os.path
import re
import sys, traceback

from switch_java_globals import PROGNAME, LOCALE_DIR
COPYRIGHT = 'Copyright (C) 2007 Red Hat, Inc.'
ssj_debug=False
none_root=False

# Internationalization.
gettext.bindtextdomain(PROGNAME, LOCALE_DIR)
gettext.textdomain(PROGNAME)
_ = gettext.gettext

TEXT_MESSAGE = _('''\
display text interface even if graphics display is available\
''')
PARSE_ERROR_MESSAGE = _('''\
An attempt to parse /var/lib/alternatives/java failed.\
''')
NO_JAVA_MESSAGE = _('''\
No supported Java packages were found.  A supported Java package is\
 one that installs a file of the form:

/usr/lib/jvm/jre-$version-$vendor/bin/java

For example, java-1.4.2-gcj-compat is a supported package because it\
 installs:

/usr/lib/jvm/jre-1.4.2-gcj/bin/java\
''')
INSTRUCTION_MESSAGE = _('Select the default Java toolset for the system.')
TITLE_MESSAGE = _('Java Toolset Configuration')
SELECTION_MESSAGE = _('Installed Java Toolsets')
ROOT_MESSAGE = _('''\
The default Java toolset can only be configured by the root user.\
''')
OK_MESSAGE = _('OK')
CLOSE_MESSAGE = _('Close')
CANCEL_MESSAGE = _('Cancel')
JAVA_PATH = '/etc/alternatives/java'

# Module-level globals

ALTERNATIVES = {}
JAVA = {}
JRE = {}
JCE = {}
JAVAC = {}
SDK = {}
PLUGIN = {}
JAVADOCDIR = {}
JAVADOCZIP = {}

class JavaOpenError(Exception):
    '''Raised if the java alternatives configuration is missing.'''
    pass
class JavaParseError(Exception):
    '''Raised if the java alternatives configuration can not be parsed.'''
    pass

def switch_java(java):
    '''Switch alternatives to the specified java.

    java is the directory suffix indicating the java name
    '''
    vendor, version, arch = get_java_split(java)
    # There are problems with the jre_ibm, jre_1.4.2, java_sdk,
    # java_sdk_1.4.2 and libjavaplugin.so alternatives in the JPackage
    # java-1.4.2-ibm and java-1.5.0-ibm packages, but not in the RHEL
    # ones.  We suppress error output from the alternatives commands
    # so that JPackage users won't be alarmed.  The only consequence
    # for them is that the seldom-used /usr/lib/jvm/jre-ibm,
    # /usr/lib/jvm/jre-1.4.2, /usr/lib/jvm/java-1.4.2 and
    # /usr/lib/jvm/java-ibm symlinks will not be updated.  In the case
    # of the plugin, JPackage and Red Hat plugin packages are
    # incompatible anyway, so failing to set an alternative will not
    # cause additional problems.
    suppress = False
    if vendor == 'ibm':
        suppress = True
    _set_alternative('java', JAVA[java])
    _set_alternative('jre_' + vendor, JRE[java], suppress)
    _set_alternative('jre_' + version, JRE[java], suppress)
    _set_alternative('jre_' + version + "_" + vendor, re.sub("/jre$", "", JRE[java].replace("java-", "jre-")), suppress)
    if JCE[java] != None:
        _set_alternative('jce_' + version + '_' + vendor + '_local_policy' + arch,
                         JCE[java])
    if JAVAC[java] != None:
        _set_alternative('javac', JAVAC[java])
        _set_alternative('java_sdk_' + vendor, SDK[java], suppress)
        _set_alternative('java_sdk_' + version, SDK[java], suppress)
        _set_alternative('java_sdk_' + version + "_" + vendor, SDK[java], suppress)
    if PLUGIN[java] != None:
        _set_alternative('libjavaplugin.so' + arch, PLUGIN[java])
    if JAVADOCDIR[java] != None:
        _set_alternative('javadocdir', JAVADOCDIR[java])
    if JAVADOCZIP[java] != None:
        _set_alternative('javadoczip', JAVADOCZIP[java])

def _set_alternative(alternative, value, suppress=False):
    '''Set the alternative to the desired value.

    Invokes /usr/sbin/alternatives --set alternative value.
    '''
    command = '/usr/sbin/alternatives --set ' + alternative + ' ' + value
    if suppress:
        command = command + ' >/dev/null 2>&1'
    _exec(command)

def _exec(commands):
    '''Execute the given string.'''
    if ssj_debug:
        print(commands)
    os.system(commands)

def version_sort_split(i):
    regex = "-+| +|\\.+"
    vendor, version, arch =  get_java_split(i)
    # group by version , vendor
    i = version + " " + vendor + " " + i
    a = re.split(regex, (i).lower())
    return a

def version_sort(i, j):
    a = version_sort_split(i)
    b = version_sort_split(j)
    la = len(a)
    lb = len(b)
    for x in range(0, min(la, lb) - 1):
        aa = a[x]
        bb = b[x]
        try:
            na = int(aa)
            nb = int(bb)
            if na < nb: return  1
            if na > nb: return -1
        except Exception:
            if aa < bb: return  1
            if aa > bb: return -1
    #return la-lb
    #its disputable how to sort jdks wit NVR and without
    return 0

def version_sort_wrapper(i, j):
    return version_sort(i[0],j[0])

def get_java_identifiers(get_alternatives_func=None):
    if get_alternatives_func is None:
        get_alternatives_func = get_alternatives
    java_identifiers = []
    # indicates whether each java_identifier is a jdk or jre
    jdks = []
    best_identifier = None
    alternatives, best = get_alternatives_func('java')
    jre_expression = re.compile('/usr/lib/jvm/jre-([^/]+)/bin/java')
    # Since Java 9, there's no separate jre dir
    jdk_expression = re.compile('/usr/lib/jvm/java-([^/]+)/(jre/)?bin/java')
    best_identifier_index = -1
    for i in range(len(alternatives)):
        alternative = alternatives[i]
        jre_search = jre_expression.search(alternative)
        jdk_search = jdk_expression.search(alternative)
        java = None
        if not jre_search == None:
            java = jre_search.group(1)
            jdks.append(False)
        elif not jdk_search == None:
            java = jdk_search.group(1)
            jdks.append(True)
        else:
            continue
        java_identifiers.append(java)
        if i == best:
            best_identifier_index = len(java_identifiers) - 1
    identifiers_and_jdks = zip(java_identifiers, jdks)
    if len(identifiers_and_jdks) > 0:
        best_identifier = identifiers_and_jdks[best_identifier_index][0]
        identifiers_and_jdks = sorted(identifiers_and_jdks, cmp=version_sort_wrapper)
        initialize_alternatives_dictionaries(identifiers_and_jdks)
        java_identifiers = sorted(java_identifiers, cmp=version_sort)
    if (ssj_debug):
        print(str(ALTERNATIVES))
        print(str(JAVA))
        print(str(JRE))
        print(str(JCE))
        print(str(JAVAC))
        print(str(SDK))
        print(str(PLUGIN))
        print(str(JAVADOCDIR))
        print(str(JAVADOCZIP))
    return java_identifiers, best_identifier

def get_plugin_alternatives(plugin_alternatives, arch):
    try:
        alternatives, best = get_alternatives('libjavaplugin.so' + arch)
        # plugin_expression = re.compile('/usr/lib/jvm/jre-([^/]*)/')
        # all fedora/rhel javas are installed into java-... dirrectory and just 
        # creates (mostly)symlinks jre-... 
        plugin_expression = re.compile('(/usr/lib/jvm/(java|jre)-([^/]*)/)|(/usr/lib.*/IcedTeaPlugin.so.*)')
        for alternative in alternatives:
            java_search = plugin_expression.search(alternative)
            if java_search == None:
                # Skip unrecognized libjavaplugin.so alternative.
                continue
            java = java_search.group(1)
            plugin_alternatives[java] = alternative
    except JavaParseError:
        # Ignore libjavaplugin.so parse errors.
        pass
    except JavaOpenError:
        # No libjavaplugin.so alternatives were found.
        pass
    return plugin_alternatives


def _get_javadocs_alternatives(key, suffix):
    javadocdir_alternatives = {}
    try:
        alternatives, best = get_alternatives(key)
        javadocdir_expression = re.compile('/usr/share/javadoc/java-([^/]*)'+suffix)
        for alternative in alternatives:
            java_search = javadocdir_expression.search(alternative)
            if java_search == None:
                # Skip unrecognized javadocdir alternative.
                continue
            java = java_search.group(1)
            javadocdir_alternatives[java] = alternative
    except JavaParseError:
        # Ignore javadocdir parse errors.
        pass
    except JavaOpenError:
        # No javadocdir alternatives were found.
        pass
    return javadocdir_alternatives

def get_javadocdir_alternatives():
    return _get_javadocs_alternatives('javadocdir', '/api')


def get_javadoczip_alternatives():
    return _get_javadocs_alternatives('javadoczip', '\.zip')

def get_alternatives(master, open_func=open, exists=os.path.exists):
    """Parse all alternatives for `master`.

    Return a tuple (alternatives, index_of_best_alternative).

    open_func is the function invoked to open the file. Defaults to
    the builtin open().

    exists is the function invoked to check that the file
    exists. Defaults to os.path.exists.

    Raise JavaOpenError if alternative file is
    not found. Raise JavaParseError if it can not be parsed.
    """
    try:
        alternative_file = open_func('/var/lib/alternatives/' + master, 'r')
    except IOError:
        raise JavaOpenError

    with alternative_file:
        return _get_alternatives(alternative_file, exists)

def _get_alternatives(alternative_file, exists):
    """Return a tuple (alternatives, index_of_best_alternative)."""
    alternatives = []
    highest_priority = -1
    best = -1
    slave_line_count = 0
    # Skip mode and master symlink lines.
    first_slave_index = 2
    index = first_slave_index
    try:
        lines = alternative_file.readlines()
        # index points to first slave line.
        line = lines[index]
        # Count number of slave lines to ignore.
        while line != '\n':
            index = index + 1
            line = lines[index]
        # index points to blank line separating slaves from target.
        slave_line_count = (index - first_slave_index) / 2
        index = index + 1
        # index points to target.
        while index < len(lines):
            line = lines[index]
            # Accept trailing blank lines at the end of the file.
            # Debian's update-alternatives requires this.
            if line == '\n':
                break
            # Remove newline.
            alternative = line[:-1]
            # Exclude alternative targets read from
            # /var/lib/alternatives/$master that do not exist in the
            # filesystem.  This inconsistent state can be the result
            # of an rpm post script failing.
            append = False
            if exists(alternative):
                append = True
                alternatives.append(alternative)
                if (ssj_debug):
                    print("found: "+str(alternative))
            index = index + 1
            # index points to priority.
            line = lines[index]
            if append:
                ss = line[:-1];
                # alternatives with --family swithch are saved as @family@priority
                if ("@" in ss):
                    splitted = ss.split("@");
                    priority = int(splitted[len(splitted)-1])
                else:
                    priority = int(ss)
                if priority > highest_priority:
                    highest_priority = priority
                    best = len(alternatives) - 1
            index = index + 1
            # index points to first slave.
            index = index + slave_line_count
            # index points to next target or end-of-file.
    except:
        if (ssj_debug):
            traceback.print_exc(file=sys.stdout)    
        raise JavaParseError
    return alternatives, best

def initialize_alternatives_dictionaries(java_identifiers_and_jdks, path_exists=None):
    if path_exists is None:
        path_exists = os.path.exists
    plugin_alternatives = get_plugin_alternatives({}, '')
    javadocdir_alternatives = get_javadocdir_alternatives()
    javadoczip_alternatives = get_javadoczip_alternatives()
    arch_found = False
    for (java, is_jdk) in java_identifiers_and_jdks:
        vendor, version, arch = get_java_split(java)
        if is_jdk:
            # since Java 9, there's no separate jre subdir
            candidate = '/usr/lib/jvm/java-' + java + '/jre/bin/java'
            if path_exists(candidate):
                JAVA[java] = candidate
            else:
                JAVA[java] = '/usr/lib/jvm/java-' + java + '/bin/java'
        else:
            JAVA[java] = '/usr/lib/jvm/jre-' + java + '/bin/java'
        # Command-to-alternative-name map to set default alternative.
        ALTERNATIVES[JAVA[java]] = java
        if is_jdk:
            # since Java 9, there's no separate jre subdir
            candidate = '/usr/lib/jvm/java-' + java + '/jre'
            if path_exists(candidate):
                JRE[java] = candidate
            else:
                JRE[java] = '/usr/lib/jvm/java-' + java
        else:
            JRE[java] = '/usr/lib/jvm/jre-' + java
        jce = '/usr/lib/jvm-private/java-' + java\
              + '/jce/vanilla/local_policy.jar'
        if path_exists(jce):
            JCE[java] = jce
        else:
            JCE[java] = None
        javac = '/usr/lib/jvm/java-' + java + '/bin/javac'
        if os.path.exists(javac):
            JAVAC[java] = javac
            SDK[java] = '/usr/lib/jvm/java-' + java
        else:
            JAVAC[java] = None
            SDK[java] = None
        if arch != '' and not arch_found:
            plugin_alternatives = get_plugin_alternatives(plugin_alternatives,
                                                          arch)
            arch_found = True
        PLUGIN[java] = None
        if java in plugin_alternatives:
            PLUGIN[java] = plugin_alternatives[java]
        for v in plugin_alternatives:
            if (str(plugin_alternatives[v]).find('IcedTeaPlugin.so')>=0 and str(java).find('openjdk')>=0):
                 PLUGIN[java]=plugin_alternatives[v]
            else:
                if (plugin_alternatives[v].find(java.split('/')[0])>=0):
                    PLUGIN[java] = plugin_alternatives[v]
        # javadoc is noarch
        archlesjava=java.replace(arch,"")
        JAVADOCDIR[java] = None
        if archlesjava in javadocdir_alternatives:
            JAVADOCDIR[java] = javadocdir_alternatives[archlesjava]
        JAVADOCZIP[java] = None
        if archlesjava in javadoczip_alternatives:
            JAVADOCZIP[java] = javadoczip_alternatives[archlesjava]

def get_default_java_command(path=JAVA_PATH):
    '''Return the default java command (default is '/etc/alternatives/java')'''
    if os.path.exists(path) and os.path.islink(path):
        return os.readlink(path)
    else:
        return None

def get_pretty_names(alternative_names):
    '''Return a dictionary that maps each item in the input list to the pretty string representation.

    See get_pretty_name for more information'''
    pretty_names = {}
    for java in alternative_names:
        pretty_names[java] = get_pretty_name(java)
    return pretty_names

def get_pretty_name(java):
    '''Return a pretty name of the form "formatted-name version optional-arch"'''
    debugPackage = False;
    if ("-debug" in java):
        debugPackage = True;
    java = java.replace("-debug", "")
    vendor, version, arch = get_java_split(java)
    if vendor in ['sun', 'blackdown', 'oracle']:
        pretty_name = vendor.capitalize() + ' ' + version
    elif vendor == 'icedtea':
        pretty_name = 'IcedTea' + ' ' + version
    elif vendor == 'openjdk':
        pretty_name = 'OpenJDK' + ' ' + version
    else:
        pretty_name = vendor.upper() + ' ' + version
    versionsplit = java.split('-')

    #handling of  legacy '1.7.0-openjdk' or '1.5.0-sun.x86_64'
    if "." not in versionsplit[len(versionsplit)-1]:
        return pretty_name
    nvr = ''
    if len(versionsplit) > 2:
        nvr = nvr + versionsplit[2]
    if len(versionsplit) > 3:
        nvr = nvr + '-' + versionsplit[3]
    if nvr == '' and len(arch) > 1:
        if arch.startswith("."):
            nvr = arch[1:]
        else:
            nvr = arch
    if (debugPackage):
        pretty_name = pretty_name + ' (' + nvr + " DEBUG"+')'
    else:
        pretty_name = pretty_name + ' (' + nvr + ')'

    return pretty_name

def get_java_split(java):
    '''Return the tuple (version, vendor, arch) for a directory suffix (1.5.0-gcj.x86_64)
    arch is empty if the directory suffix does not contain arch.'''
    if ("-debug" in java):
        java=java.replace("-debug","")
    vendor_arch = java.split('-')
    if "." not in vendor_arch[len(vendor_arch)-1]:
        return vendor_arch[1], vendor_arch[0], ''
    vendor_arch = java.split('-')[1].split('.')
    vendor = vendor_arch[0]
    arch = ''
    #path can be full of dots - as in jdk>6 full version is included
    vendor_version_arch = java.split('.')
    if len(vendor_version_arch) > 1:
        arch = '.' + vendor_version_arch[len(vendor_version_arch)-1]
    version = java.split('-')[0]
    return vendor, version, arch
